OsmAnd/src/net/osmand/plus/views/layers/MapTileLayer.java

```diff
@@ -232,25 +244,86 @@ public MapTileAdapter getMapTileAdapter() {
 		return null;
 	}
 
+	private void setLayerProvider(ITileSource mapSrc) {
+		if (view == null || mapSrc == null) {
+			return;
+		}
+
+		final MapRendererView mapRenderer = view.getMapRenderer();
+		if (mapRenderer != null) {
+			TileSourceProxyProvider prov = new TileSourceProxyProvider(view.getApplication(), mapSrc);
+			mapRenderer.setMapLayerProvider(view.getLayerIndex(this), prov.instantiateProxy(true));
+			prov.swigReleaseOwnership();
+		}
+	}
+
+	private void resetLayerProvider() {
+		if (view == null) {
+			return;
+		}
+
+		final MapRendererView mapRenderer = view.getMapRenderer();
+		if (mapRenderer != null) {
+			mapRenderer.resetMapLayerProvider(view.getLayerIndex(this));
+		}
+	}
+
+	private void updateAlpha() {
+		if (view != null) {
+			int alpha = this.getAlpha();
+
+			final MapRendererView mapRenderer = view.getMapRenderer();
+			if (mapRenderer != null) {
+				MapLayerConfiguration mapLayerConfiguration = new MapLayerConfiguration();
+				mapLayerConfiguration.setOpacityFactor(((float) alpha) / 255.0f);
+				mapRenderer.setMapLayerConfiguration(view.getLayerIndex(this), mapLayerConfiguration);
+			}
+		}
+	}
+
 	@SuppressLint("WrongCall")
 	@Override
 	public void onPrepareBufferImage(Canvas canvas, RotatedTileBox tileBox, DrawSettings drawSettings) {
-		if ((map == null && mapTileAdapter == null) || !visible) {
```

Comment:
We have to perform any updates of rendering parameters in the scope of this method (`onPrepareBufferImage`) according to current logic, thus specifically in this case there are no reason to check `mapTileAdapter` at this place (it will be checked by few lines bellow) because we have to processed layer visibility firstly.